genienlp/data_utils/bootleg.py

```diff
@@ -0,0 +1,381 @@
+import os
+import ujson
+import numpy as np
+import logging
+
+from .database_utils import is_banned
+
+from bootleg.extract_mentions import extract_mentions
+from bootleg.utils.parser_utils import get_full_config
+from bootleg import run
+
+
+logger = logging.getLogger(__name__)
+
+
+def reverse_bisect_left(a, x, lo=None, hi=None):
+    """Find item x in list a, and keep it reverse-sorted assuming a
+    is reverse-sorted.
+    """
+    if lo is None:
+        lo = 0
+    if hi is None:
+        hi = len(a)
+    while lo < hi:
+        mid = (lo + hi) // 2
+        if x > a[mid]:
+            hi = mid
+        else:
+            lo = mid + 1
+    return lo
+
+
+def post_process_bootleg_types(qid, type, title, almond_domains):
+    # TODO if training on multiple domains (in one run) these mapping should be modified
+    # e.g. song is mapped to book which is not correct if training on music domain too
+    for domain in almond_domains:
+        if domain == 'restaurants':
+            # pizzeria
+            if qid == 'Q177':
+                type = 'Q571'
+            
+            # ramen
+            if qid == 'Q1051265':
+                type = 'Q1778821'
+            
+            if 'cuisine' in title or 'pasta' in title or 'culture of ' in title or \
+                    title in ['food', 'type of food or dish', 'dish', 'convenience food', 'rice dish',
+                              'dish', 'food ingredient', 'stuffed pasta', 'raw fish dish',
+                              'soup', 'country', 'sovereign state', 'noodle', 'intangible cultural heritage']:
+                type = 'Q1778821'
+            
+            elif title in ['city of the United States', 'big city', 'city with millions of inhabitants',
+                           'commune of France']:
+                type = 'Q2221906'
+            
+            elif 'restaurant chain' in title or title in ['restaurant', 'food manufacturer']:
+                type = 'Q571'
+            
+            elif 'writer' in title:
+                type = 'Q5'
+            
+            elif title in ['musical group', 'Wikimedia disambiguation page', 'Wikimedia list article', 'film']:
+                type = 'unk'
```

Comment:
Why is `type` a QID and not a ThingTalk type?

Why do you not traverse the wikidata graph up until you hit the right type instead of writing a bunch of if statements?

Reply:
That was the older implementation which was slower and had lower results for Spotify...

Reply:
Which part was the older implementation, traversing the graph? Did you investigate why?

Reply:
Yeah traversing. The dictionary that maps Qids to types is incomplete. Bootleg uses some heuristics such as comparing the frequency of a term appearing in Wikipedia to how often it appears as a labeled mention. If it appears as a mention less than <~3% of the time, they filter it out. Thus traversing doesn't always reach the type we want.

Reply:
But we have the real Wikidata in Elastic Search. We can do a round of processing on the Bootleg list of types and map them to a ThingTalk type ahead of time with some ES searches. In fact, we only need to do this once and store the processed types, as long as the mapping of "root QID" to ThingTalk type doesn't change.

Reply:
Yeah, we can use our ES to complete Bootleg databases. Will add it to the next PR's bucket list.